File: src/go/problems/IsAnagram_test.go
// https://leetcode.com/problems/valid-anagram/
package problems

import "testing"

func TestIsAnagram(t *testing.T) {
	result := IsAnagram("anagram", "nagaram")
	t.Log(result)
}

func IsAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	set := map[byte]int{}
	length := len(s)
	for i := 0; i < length; i++ {
		set[s[i]]++
		set[t[i]]--
	}

	for _, v := range set {
		if v != 0 {
			return false
		}
	}

	return true
}

File: src/python/problems/is_anagram_test.py
# https://leetcode.com/problems/valid-anagram

from core.problem_base import *

class IsAnagram(ProblemBase):
    def Solution(self, s: str, t: str) -> bool:
        if len(s) != len(t):
            return False

        map = {}
        for i in range(len(s)):
            map[s[i]] = map.get(s[i], 0) + 1
            map[t[i]] = map.get(t[i], 0) - 1

        for _, v in map.items():
            if v != 0:
                return False

        return True

if __name__ == '__main__':
    TestGen(IsAnagram) \
        .Add(lambda tc: tc.Param("anagram").Param("nagaram").Result(True)) \
        .Add(lambda tc: tc.Param("rat").Param("cat").Result(False)) \
        .Run()
